### ISSUE TITLE
scylla_raid_setup: failed due to the "mdmonitor.service: Failed with result 'protocol'"

### ISSUE THREAD

**Issue by dkropachev** [scylladb/scylla#9540]
Occasionally `scylla_raid_setup` fails due to the `mdmonitor.service` is unable to start.

Log if `scylla_setup` call:
```
sudo /usr/lib/scylla/scylla_setup --nic eth0 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '
Exit code: 1
Stdout:
=                       reflink=1
data     =                       bsize=4096   blocks=294812928, imaxpct=5
=                       sunit=256    swidth=768 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=143951, version=2
=                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
/dev/md0 will be used to setup a RAID
Creating RAID0 for scylla using 3 disk(s): /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3
RAID setup failed.
Stderr:
Job for mdmonitor.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status mdmonitor.service" and "journalctl -xe" for details.
Traceback (most recent call last):
File "/opt/scylladb/scripts/libexec/scylla_raid_setup", line 188, in <module>
md_service.start()
File "/opt/scylladb/scripts/scylla_util.py", line 790, in start
return run('systemctl {} start {}'.format(self.ctlparam, self._unit), shell=True, check=True)
File "/opt/scylladb/python3/lib64/python3.9/subprocess.py", line 524, in run
raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'systemctl  start mdmonitor.service' returned non-zero exit status 1.
Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 3491, in node_setup
cl_inst.node_setup(_node, **setup_kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 4201, in node_setup
self._scylla_post_install(node, install_scylla, nic_devname)
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 4277, in _scylla_post_install
node.scylla_setup(disks, devname)
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 113, in inner
res = func(*args, **kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2279, in scylla_setup
self.remoter.run('sudo /usr/lib/scylla/scylla_setup --nic {} --disks {} --no-io-setup {}'
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run
result = _run()
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 64, in inner
return func(*args, **kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run
return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute
result = connection.run(**command_kwargs)
File "<decorator-gen-3>", line 2, in run
File "/usr/local/lib/python3.10/site-packages/fabric/connection.py", line 30, in opens
return method(self, *args, **kwargs)
File "/usr/local/lib/python3.10/site-packages/fabric/connection.py", line 723, in run
return self._run(self._remote_runner(), command, **kwargs)
File "/usr/local/lib/python3.10/site-packages/invoke/context.py", line 102, in _run
return runner.run(command, **kwargs)
File "/usr/local/lib/python3.10/site-packages/invoke/runners.py", line 380, in run
return self._run_body(command, **kwargs)
File "/usr/local/lib/python3.10/site-packages/invoke/runners.py", line 442, in _run_body
return self.make_promise() if self._asynchronous else self._finish()
File "/usr/local/lib/python3.10/site-packages/invoke/runners.py", line 509, in _finish
raise UnexpectedExit(result)
invoke.exceptions.UnexpectedExit: Encountered a bad command exit code!
Command: 'sudo /usr/lib/scylla/scylla_setup --nic eth0 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '
Exit code: 1
Stdout:
=                       reflink=1
data     =                       bsize=4096   blocks=294812928, imaxpct=5
=                       sunit=256    swidth=768 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=143951, version=2
=                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
/dev/md0 will be used to setup a RAID
Creating RAID0 for scylla using 3 disk(s): /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3
RAID setup failed.
Stderr:
Job for mdmonitor.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status mdmonitor.service" and "journalctl -xe" for details.
Traceback (most recent call last):
File "/opt/scylladb/scripts/libexec/scylla_raid_setup", line 188, in <module>
md_service.start()
File "/opt/scylladb/scripts/scylla_util.py", line 790, in start
return run('systemctl {} start {}'.format(self.ctlparam, self._unit), shell=True, check=True)
File "/opt/scylladb/python3/lib64/python3.9/subprocess.py", line 524, in run
raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'systemctl  start mdmonitor.service' returned non-zero exit status 1.
```

System.log :

```
2021-10-22T05:10:03+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !INFO    | kernel: md0: detected capacity change from 0 to 1207553753088
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !INFO    | systemd[1]: Reloading.
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !INFO    | systemd[1]: Reloading.
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !INFO    | systemd[1]: Starting Software RAID monitoring and management...
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !INFO    | systemd[1]: mdmonitor.service: New main PID 54588 does not exist or is a zombie.
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !WARNING | systemd[1]: mdmonitor.service: Failed with result 'protocol'.
2021-10-22T05:10:04+00:00  rolling-upgrade-master-centos-8-db-node-1f1eaf0e-0-2 !ERR     | systemd[1]: Failed to start Software RAID monitoring and management.
```

*Installation details*
Scylla version (or git commit hash): 4.5.0-0.20211006.47be33a10 with build-id cae2146dc91480331365938fddb866fe4ecc0c56
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): Centos 8

Platform (physical/VM/cloud instance type/docker): GCE
Disks: (SSD/HDD, count): 3 nvme
Image: http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-10-20T08:48:46Z/scylla.repo

*Logs*
https://cloudius-jenkins-test.s3.amazonaws.com/1f1eaf0e-a3c5-4a50-bceb-34b25c7b18a5/20211022_051038/db-cluster-1f1eaf0e.tar.gz

**Comment by fgelcer** [comment]
my test is failing when starting from 4.6:
```
Test details
Test: upgrade_test.UpgradeTest.test_rolling_upgrade
Build number: 2
Backend: gce: eu-west-1
Kernel version: 4.18.0-305.25.1.el8_4.x86_64
Test-id: 19f47e44-bd67-4349-8ec1-02bb3901353b
Start time: 2021-12-05 20:34:00
End time: 2021-12-05 20:43:07
Started by user: fabio
Test result
FAILED
System under test
ScyllaDB version: 4.6.rc0-0.20211130.afc18d507 with build-id 4b0d4e8d54151969f94c49c4aaf6c34c6ef5136b (-)
Target ScyllaDB repo: http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-11-28T03:59:37Z/scylla.repo
Instance type: n1-highmem-8
Number of ScyllaDB nodes: 4
```

**Comment by syuu1228** [comment]
I tried to reproduce the issue on my local Vagrant CentOS8 instance with 2 additional disks for RAID, but it just succeeded.

```
[vagrant@localhost ~]$ rpm -qa|grep scylla
scylla-conf-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-kernel-conf-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-tools-core-4.7.dev-0.20211128.44f4ea38c.noarch
scylla-server-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-node-exporter-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-tools-4.7.dev-0.20211128.44f4ea38c.noarch
scylla-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-python3-4.7.dev-0.20211128.44f4ea38c.x86_64
scylla-jmx-4.7.dev-0.20211128.44f4ea38c.noarch
```
```
Do you want to setup RAID and XFS?
It is recommended to use RAID and XFS for Scylla data. If you select yes, you will be prompted to choose the unmounted disks to use for Scylla data. Selected disks are formatted as part of the process.
Yes - choose a disk/disks to format and setup for RAID and XFS. No - skip this step.
[YES/no]
Are you sure you want to setup RAID and XFS?
If you choose Yes, the selected drive will be reformated, erasing all existing data in the process.
[YES/no]
Do you want to change RAID level to RAID5?
If you choose Yes, change RAID level to RAID5. Otherwise we will use RAID0.
[yes/NO]  
Enable XFS online discard?
The default (Yes) asks the disk to recycle SSD cells as soon as files are deleted. 4.18+ kernel recommended for this option.
[YES/no]   
Please select unmounted disks from the following list: /dev/sdb, /dev/sdc
type 'cancel' to cancel RAID/XFS setup.
type 'done' to finish selection. Selected: 
> /dev/sdb
type 'cancel' to cancel RAID/XFS setup.
type 'done' to finish selection. Selected: /dev/sdb
> /dev/sdc
type 'cancel' to cancel RAID/XFS setup.
type 'done' to finish selection. Selected: /dev/sdb, /dev/sdc
> done
/dev/md0 will be used to setup a RAID
Last metadata expiration check: 0:02:16 ago on Mon 06 Dec 2021 09:59:25 AM UTC.
Dependencies resolved.
================================================================================
 Package         Arch             Version                Repository        Size
================================================================================
Installing:
 mdadm           x86_64           4.2-rc2.el8            BaseOS           460 k

Transaction Summary
================================================================================
Install  1 Package

Total download size: 460 k
Installed size: 1.1 M
Downloading Packages:
mdadm-4.2-rc2.el8.x86_64.rpm                    379 kB/s | 460 kB     00:01    
--------------------------------------------------------------------------------
Total                                           189 kB/s | 460 kB     00:02     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                        1/1 
  Installing       : mdadm-4.2-rc2.el8.x86_64                               1/1 
  Running scriptlet: mdadm-4.2-rc2.el8.x86_64                               1/1 
  Verifying        : mdadm-4.2-rc2.el8.x86_64                               1/1 

Installed:
  mdadm-4.2-rc2.el8.x86_64                                                      

Complete!
Creating fRAID{args.raid_level} for scylla using 2 disk(s): /dev/sdb,/dev/sdc
mdadm: Defaulting to version 1.2 metadata
mdadm: array /dev/md0 started.
meta-data=/dev/md0               isize=512    agcount=16, agsize=2618368 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1
data     =                       bsize=1024   blocks=41893888, imaxpct=25
         =                       sunit=1024   swidth=2048 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=1024   blocks=20480, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
Created symlink /etc/systemd/system/multi-user.target.wants/var-lib-scylla.mount → /etc/systemd/system/var-lib-scylla.mount.
Do you want to enable coredumps?
Yes - sets up coredump to allow a post-mortem analysis of the Scylla state just prior to a crash. No - skips this step.
[YES/no]
```
```
[vagrant@localhost ~]$ systemctl status mdmonitor.service
● mdmonitor.service - Software RAID monitoring and management
   Loaded: loaded (/usr/lib/systemd/system/mdmonitor.service; enabled; vendor p>
   Active: inactive (dead) since Mon 2021-12-06 10:01:47 UTC; 2min 50s ago
 Main PID: 22246 (code=exited, status=0/SUCCESS)
```

@dkropachev Please send AMI ID, instance type, and Scylla repository URL.

**Comment by dkropachev** [comment]
> @dkropachev Please send AMI ID, instance type, and Scylla repository URL.

Sorry for confision, It is not and AWS, but GCP rather http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-10-20T08:48:46Z/scylla.repo

**Comment by slivne** [comment]
We need the system logs on such failures to analyze what may have been the issue 

@fgelcer / @dkropachev 

Lets retain an instance this happen on - so that @syuu1228 will login and be able to check that

**Comment by fgelcer** [comment]
@slivne , i can try to re-run few times the rolling upgrade tests on master to see if it reproduces, i'm just not sure if the nodes are kept in case of a failure in the node's setup...

**Comment by fgelcer** [comment]
> We need the system logs on such failures to analyze what may have been the issue
> 
> @fgelcer / @dkropachev
> 
> Lets retain an instance this happen on - so that @syuu1228 will login and be able to check that

reproduced, node is alive `rolling-upgrade-upgrade--centos-8-db-node-a671a386-0-1 [35.196.47.124 | 10.142.0.29]`
@syuu1228 FYI

**Comment by syuu1228** [comment]
Okay, I finally able to reproduce this issue.
It looks like timing bug, I needed retry more than 5 times, as @fgelcer commented.

```
[syuu@takuya-test-5 yum.repos.d]$ sudo /usr/lib/scylla/scylla_setup --nic eth0 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory /
WARN  2021-12-11 05:21:06,138 [shard 0] iotune - Available space on filesystem at /var/tmp/mnt: 121 MB: is less than recommended: 10 GB
INFO  2021-12-11 05:21:06,138 [shard 0] iotune - /var/tmp/mnt passed sanity checks
This is a supported kernel version.
Created symlink /etc/systemd/system/multi-user.target.wants/scylla-server.service → /usr/lib/systemd/system/scylla-server.service.
ntp is already configured, skip setup
/dev/md0 will be used to setup a RAID
Creating fRAID{args.raid_level} for scylla using 3 disk(s): /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3
mdadm: Defaulting to version 1.2 metadata
mdadm: array /dev/md0 started.
meta-data=/dev/md0               isize=512    agcount=33, agsize=9212672 blks
         =                       sectsz=4096  attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1
data     =                       bsize=4096   blocks=294812928, imaxpct=5
         =                       sunit=256    swidth=768 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=143951, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
Job for mdmonitor.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status mdmonitor.service" and "journalctl -xe" for details.
Traceback (most recent call last):
  File "/opt/scylladb/scripts/libexec/scylla_raid_setup", line 198, in <module>
    md_service.start()
  File "/opt/scylladb/scripts/scylla_util.py", line 1010, in start
    return run('systemctl {} start {}'.format(self.ctlparam, self._unit), shell=True, check=True)
  File "/opt/scylladb/python3/lib64/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'systemctl  start mdmonitor.service' returned non-zero exit status 1.
RAID setup failed.
```

**Comment by syuu1228** [comment]
I realized that mdmonitor.service on CentOS8 is actually ALWAYS failing to start as daemon, the only difference is `systemctl status mdmonitor.service` result.
The process immediately shutdown right after started, without any error and returns 0.
The behavior is similar to --oneshot mode, maybe it is running in the mode even the option not specified.
The issue only occurs on mdadm-4.2-rc2.el8.x86_64, maybe it's a bug on 4.2-rc2.
I also able to reproduce same issue on Fedora 35 (mdadm-4.2-rc2.fc35.x86_64).

The issue is not occurs on mdadm-4.1-15.el8.x86_64 and older, possible workaround is downgrade the package to mdadm-4.1.

Both CentOS7 and Ubuntu 20.04 don't have this issue since they are using mdadm-4.1.

**Comment by syuu1228** [comment]
Another possible solution is, just stop using mdmonitor.service.
When I added this on scylla_raid_setup, I was thinking the service is needed to enable RAID volume, but it actually does nothing to enable RAID, it just poll RAID status and send email to root account when the status changed (RAID volume degraded, etc).
Since it just sending root account instead of e-mail address, most of users won't read it anyway.

**Comment by avikivity** [comment]
It also writes to syslog. I agree it's low-value, we can consider disabling it. RAID-0 failure is fatal so the admin will be aware of it.

For RAID-5 it's different, we need some notification about disk failure.

**Comment by syuu1228** [comment]
More note: The issue reproduced at GCE, local Vagrant (VirtualBox), and local baremetal machine. So it doesn't looks like specific enviroment problem.

**Comment by syuu1228** [comment]
I checked difference between mdadm-4.1-14.el8.x86_64 and mdadm-4.1-15.el8.x86_64.
diff of mdadm.spec: https://gist.github.com/syuu1228/70aea940e3e100e3bc2eee31fc4d13f9

And I found this patch causes the issue:
0093-Monitor-stop-notifing-about-containers.patch
```
From 007087d0898a045901e4e120296e6d9b845b20a6 Mon Sep 17 00:00:00 2001
From: Mariusz Tkaczyk <mariusz.tkaczyk@intel.com>
Date: Wed, 9 Sep 2020 10:31:18 +0200
Subject: [PATCH 093/108] Monitor: stop notifing about containers.

Stop reporting any events from container but still track them,
it is important for spare migration.
Stop mdmonitor if no redundant array is presented in mdstat.
There is nothing to follow.

Signed-off-by: Mariusz Tkaczyk <mariusz.tkaczyk@intel.com>
Signed-off-by: Jes Sorensen <jsorensen@fb.com>
---
 Monitor.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/Monitor.c b/Monitor.c
index 80a3200..aed7a69 100644
--- a/Monitor.c
+++ b/Monitor.c
@@ -212,15 +212,24 @@ int Monitor(struct mddev_dev *devlist,
 		int new_found = 0;
 		struct state *st, **stp;
 		int anydegraded = 0;
+		int anyredundant = 0;
 
 		if (mdstat)
 			free_mdstat(mdstat);
 		mdstat = mdstat_read(oneshot ? 0 : 1, 0);
 
-		for (st = statelist; st; st = st->next)
+		for (st = statelist; st; st = st->next) {
 			if (check_array(st, mdstat, c->test, &info,
 					increments, c->prefer))
 				anydegraded = 1;
+			/* for external arrays, metadata is filled for
+			 * containers only
+			 */
+			if (st->metadata && st->metadata->ss->external)
+				continue;
+			if (st->err == 0 && !anyredundant)
+				anyredundant = 1;
+		}
 
 		/* now check if there are any new devices found in mdstat */
 		if (c->scan)
@@ -236,6 +245,9 @@ int Monitor(struct mddev_dev *devlist,
 		if (!new_found) {
 			if (oneshot)
 				break;
+			else if (!anyredundant) {
+				break;
+			}
 			else {
 				mdstat_wait(c->delay);
 				mdstat_close();
@@ -542,7 +554,8 @@ static int check_array(struct state *st, struct mdstat_ent *mdstat,
 		st->err = 0;
 		st->percent = RESYNC_NONE;
 		new_array = 1;
-		alert("NewArray", st->devname, NULL, ainfo);
+		if (!is_container)
+			alert("NewArray", st->devname, NULL, ainfo);
 	}
 
 	if (st->utime == array.utime && st->failed == sra->array.failed_disks &&
@@ -676,7 +689,7 @@ static int check_array(struct state *st, struct mdstat_ent *mdstat,
 	return retval;
 
  disappeared:
-	if (!st->err)
+	if (!st->err && !is_container)
 		alert("DeviceDisappeared", dev, NULL, ainfo);
 	st->err++;
 	goto out;
-- 
2.7.5
```

But this patch itself seems correct, I think real issue is `check_array()` returns `st->err == 1`.
I checked `check_array()` code, looks like we hit this code:
```
	/* It's much easier to list what array levels can't
	 * have a device disappear than all of them that can
	 */
	if (sra->array.level == 0 || sra->array.level == -1) {
		if (!st->err && !st->from_config)
			alert("DeviceDisappeared", dev, " Wrong-Level", ainfo);
		st->err++;
		goto out;
	}
``
The code looks like unconditionally set st->err++ when RAID level is 0.
And the patch I pasted avobe will set anyredundant = 0 when st->err is not 0, and it will break Monitor loop `if (!anyredundant)`.
So it will unconditionally exit monitor mode when RAID0.

**Comment by syuu1228** [comment]
I confirmed this issue does not occur on RAID5 configuration, so this is RAID0 only problem as I described avobe.

@avikivity How about to skip enabling mdmonitor.service when RAID0, since we only need it for RAID5 and the error only occur on RAID0?

**Comment by avikivity** [comment]
Let's file a bug. They might fix it quickly and so we won't have to do anything.

**Comment by syuu1228** [comment]
It also reproducible on RHEL8 not just CentOS8, reported bug as RHEL8 bug.
https://bugzilla.redhat.com/show_bug.cgi?id=2031936

I couldn't find issue tracker for mdadm, posted it at Red Hat Bugzilla.

**Comment by aleksbykov** [comment]
Same issue happened during upgrade from 4.3 -> 4.4.8. During configuring node for 4.3
in job: https://jenkins.scylladb.com/job/scylla-4.4/job/rolling-upgrade/job/rolling-upgrade-centos8-test/89/console
error:
```
sudo /usr/lib/scylla/scylla_setup --nic eth0 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '
18:53:09  
18:53:09  Exit code: 1
18:53:09  
18:53:09  Stdout:
18:53:09  
18:53:09  data     =                       bsize=4096   blocks=294812928, imaxpct=5
18:53:09           =                       sunit=256    swidth=768 blks
18:53:09  naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
18:53:09  log      =internal log           bsize=4096   blocks=143951, version=2
18:53:09           =                       sectsz=4096  sunit=1 blks, lazy-count=1
18:53:09  realtime =none                   extsz=4096   blocks=0, rtextents=0
18:53:09  /dev/md0 will be used to setup a RAID
18:53:09  Creating RAID0 for scylla using 3 disk(s): /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3
18:53:09  You current Scylla release is 4.3.6 the latest minor release is 4.5.2 go to http://www.scylladb.com for upgrade instructions
18:53:09  RAID setup failed.
18:53:09  
18:53:09  Stderr:
18:53:09  
18:53:09  Traceback (most recent call last):
18:53:09    File "/opt/scylladb/scripts/libexec/scylla_raid_setup", line 184, in <module>
18:53:09      md_service.start()
18:53:09    File "/opt/scylladb/scripts/scylla_util.py", line 765, in start
18:53:09      return run('systemctl {} start {}'.format(self.ctlparam, self._unit))
18:53:09    File "/opt/scylladb/scripts/scylla_util.py", line 513, in run
18:53:09      return subprocess.run(cmd, stdout=stdout, stderr=stderr, shell=shell, check=exception, env=scylla_env).returncode
18:53:09    File "/opt/scylladb/python3/lib64/python3.8/subprocess.py", line 512, in run
18:53:09      raise CalledProcessError(retcode, process.args,
18:53:09  subprocess.CalledProcessError: Command '['systemctl', 'start', 'mdmonitor.service']' returned non-zero exit status 1.

```

**Comment by syuu1228** [comment]
@avikivity Isn't it better to merge workaround patch(#9782) ? Looks like the bug still not fixed.

**Comment by syuu1228** [comment]
@avikivity ping

**Comment by avikivity** [comment]
Backported to 4.4, 4.5, 4.6.

**Comment by avikivity** [comment]
> I confirmed this issue does not occur on RAID5 configuration, so this is RAID0 only problem as I described avobe.

On the bug tracker they say that's expected behavior.

> 
> @avikivity How about to skip enabling mdmonitor.service when RAID0, since we only need it for RAID5 and the error only occur on RAID0?


So this is the correct fix.

**Comment by syuu1228** [comment]
> > I confirmed this issue does not occur on RAID5 configuration, so this is RAID0 only problem as I described avobe.
> 
> On the bug tracker they say that's expected behavior.
> 
> > @avikivity How about to skip enabling mdmonitor.service when RAID0, since we only need it for RAID5 and the error only occur on RAID0?
> 
> So this is the correct fix.

Okay.. I will send new patch to implement it.

**Comment by fgelcer** [comment]
i'm not sure this issue was resolved in 4.3, but in the rolling upgrade job of 2021.1 (when starting from a 4.3 DB node), we are seeing the following error:
```
Traceback (most recent call last):
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/tester.py", line 121, in wrapper
    return method(*args, **kwargs)
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/utils/decorators.py", line 110, in inner
    res = func(*args, **kwargs)
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/tester.py", line 490, in setUp
    self.init_nodes(db_cluster=self.db_cluster)
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/tester.py", line 441, in init_nodes
    db_cluster.wait_for_init(node_list=db_cluster.nodes)
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/cluster.py", line 3562, in wrapper
    verify_node_setup(start_time)
  File "/jenkins/slave/workspace/enterprise-2021.1/rolling-upgrade/rolling-upgrade-centos8-test/scylla-cluster-tests/sdcm/cluster.py", line 3536, in verify_node_setup
    raise NodeSetupFailed(node=node, error_msg=setup_exception[0], traceback_str=setup_exception[1])
sdcm.cluster.NodeSetupFailed: [Node rolling-upgrade-2021-1-centos-db-node-a0499866-0-1 [35.243.224.188 | 10.142.0.166] (seed: True)] NodeSetupFailed: Encountered a bad command exit code!

Command: 'sudo /usr/lib/scylla/scylla_setup --nic eth0 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '

Exit code: 1

Stdout:

data     =                       bsize=4096   blocks=294812928, imaxpct=5
         =                       sunit=256    swidth=768 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=143951, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
/dev/md0 will be used to setup a RAID
Creating RAID0 for scylla using 3 disk(s): /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3
You current Scylla release is 4.3.6 the latest minor release is 4.5.3 go to http://www.scylladb.com for upgrade instructions
RAID setup failed.

Stderr:

Traceback (most recent call last):
  File "/opt/scylladb/scripts/libexec/scylla_raid_setup", line 184, in <module>
    md_service.start()
  File "/opt/scylladb/scripts/scylla_util.py", line 765, in start
    return run('systemctl {} start {}'.format(self.ctlparam, self._unit))
  File "/opt/scylladb/scripts/scylla_util.py", line 513, in run
    return subprocess.run(cmd, stdout=stdout, stderr=stderr, shell=shell, check=exception, env=scylla_env).returncode
  File "/opt/scylladb/python3/lib64/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['systemctl', 'start', 'mdmonitor.service']' returned non-zero exit status 1.
```
